### DISCUSSION TITLE
Dynamic context in Middleware

### DISCUSSION THREAD

**Discussion by brandonwestcott** [prisma/prisma#13366] [Q&A]
## Problem
**Middleware lacks the ability to receive dynamic context.**

This problem has been captured a number of times by the community already. The typically covered use cases are per request query filtering (user or multitenant), audit trails, and improved logging.

https://github.com/prisma/prisma/issues/12757
https://github.com/prisma/prisma/issues/10950
https://github.com/prisma/prisma/issues/10895
https://github.com/prisma/prisma/issues/9879
https://github.com/prisma/prisma/pull/9759#issuecomment-1112408978
https://github.com/prisma/prisma/issues/7596
https://github.com/prisma/prisma/issues/6882
https://github.com/prisma/prisma/issues/5923
https://github.com/prisma/prisma/issues/4217
https://github.com/prisma/prisma/issues/1902

## Suggested solution
Mainly this is just a request for the prisma team to provide guidance on what the suggest path forward is.
- Is solving this problem on the roadmap?
- Should we be [extending the prisma client](https://github.com/prisma/prisma/issues/7161)? <- though has its limitations, especially w/r/t logging
- Should we be hacking together per thread middleware using [continuation local storage](https://github.com/prisma/prisma/pull/9759#issuecomment-1115916461)? <- again, limitations on complex cases

A PR is already inflight, but has not received much attention from the team:
https://github.com/prisma/prisma/pull/9759

Overall thanks prisma team for the great work. I've been happily using prisma in production on two different products for nearly two years. The two biggest thorns have been the lack of read then update / code driven transcations (already resolved with the callback, thanks!) and this. 1 down, 1 to go.

**Comment by faoiseamh** [comment]
I'll second the need for this. Some direction from the core team would be greatly appreciated. Our use-case involves database(-middleware)-level audit logging which, so far, we are planning to achieve with a context field per-table, however this requires modifying all of our tables and it's not a great solution. Functions are not a viable solution for this as we want to wrap every potential query with audit logging. We'd like to achieve this by passing the context in some sort of variable, perhaps one we inject in an extended client or something along those lines?

**Comment by brandonwestcott** [comment]
@janpio does the prisma team have any insights or recommendations here? Our team is planning out the next version of our api stack and this is a big factor in analyzing our cost of sticking with prisma.

**Reply by faoiseamh** [reply]
In the next few weeks we'll be experimenting with extending the prisma client to add a method to inject controller / request context that will then be accessible in middleware. For this to work with concurrent requests, we're also planning to create a pool of prisma clients that are first mutex-locked before being used in a request, ensuring that client is not used by any other requests for the duration of the request and thus the client-wide dynamic context is correct and not overwritten by some other request. Once the request completes, we will release the client back to the pool for use in other requests. Granted we haven't tested any of this yet, but given my cursory understanding of Prisma, Express, and Node, this is the best aproa

**Comment by tinezmatias** [comment]
I continued with the work of this [9759](https://github.com/prisma/prisma/pull/9759), in this other [14193](https://github.com/prisma/prisma/pull/14193)

**Comment by supalarry** [comment]
Would love if the information from the dynamic context would also be logged by the query engine so that the following is possible

```
prisma.$on('query', (e) => {
  console.log('Query: ' + e.query)
  console.log('Metadata ' + e.meta) // the dynamic context
})
```

Our team wants to catch slow queries and inform the developer where in codebase did the slow query occur.

**Reply by janpio** [reply]
That sounds like a valid feature request. Please open an issue. Thanks.

**Comment by psikoi** [comment]
@matthewmueller 's [proposal](https://github.com/prisma/prisma/issues/6882#issuecomment-1175427037) in #6882 seems to be a great solution to many of the use cases people have reported.

**Comment by rijk** [comment]
Would definitely nice to have something built-in functionality for this. Middleware is the ideal place to do permission checks for data mutations, filtering etc. I currently rolled my own solution, some snippets:

```typescript
interface AuthenticatedPrismaClient extends PrismaClient {
  authenticate: (session: Session | null) => AuthenticatedPrismaClient
  userId?: string
  organizationId?: string
  roles?: Set<string>
}

// Used for checking roles in rules middleware
const authenticate = (session: Session | null) => {
  prisma.roles = new Set(session?.roles)
  prisma.userId = session?.user.id
  prisma.organizationId = session?.organizationId
  return prisma
}

export const prisma: AuthenticatedPrismaClient = Object.assign(new PrismaClient(), { authenticate })
prisma.$use(filter)
prisma.$use(rules)
```

Example of the middleware:

```typescript
export const filter: Prisma.Middleware = async (params, next) => {
  if (params.action === 'findMany' && params.model === 'User') {
    params.args.where ??= {}
    params.args.where['organizationId'] = prisma.organizationId
  }

  return next(params)
}
```

```typescript
const notAllowedError = new Error('Sorry, you are not allowed to do this')

export const rules: Prisma.Middleware = async (params, next) => {
  if (
    params.model === 'Path' ||
    params.model === 'Module' ||
    params.model === 'Lesson'
  ) {
    if (/update/.test(params.action) || /create/.test(params.action)) {
      if (!prisma.roles?.has('system-admin')) throw notAllowedError
    }

    // check field permissions, etc
  }

  return next(params)
}
```

Then I can use it like this in my (Next.js React Server Component) page:

```tsx

export default async function AdminPage() {
  const session = await getServerSession()
  const users = await prisma.authenticate(session).user.findMany()

  // etc
}
```

Obviously, this API is highly targeted at my app and use case. But I think a `client.context(whatever)` API as @ntgussoni suggested [here](https://github.com/prisma/prisma/issues/5923#issuecomment-850292674) could def work for this. That would also open up possibilities to create a generic community middleware for e.g. checking field permissions, similar to `prisma-redis-middleware`.

**Reply by kole** [reply]
This was incredibly helpful, thanks for posting!

**Reply by ali-shabani** [reply]
This is awesome, but this has some flaws if you are considering this as adding context to the middleware. For example, you will get the wrong context if you resolve multiple Prisma queries without waiting for the previous one to resolve.

```javascript
Promise.all([
  prisma.setContext({id: 1}).model.findMany(),
  prisma.setContext({id: 2}).model.findMany()
])

// the context id in the middleware for the above operations will be {id: 2}
```

**Reply by rijk** [reply]
That's a good point, although I struggle to imagine a use case where this would happen?

Also, if there's a built in .context function, you would more likely call it once, before running the queries. That would remove that gotcha.

**Reply by ali-shabani** [reply]
My main point was that you need to be aware of your code flow, for example:
```javascript
app.post('/test', (req, res) => {
  db.setContext({user: req.user}).model.create({}).then() // let's make it non-blocking

  res.send('ok')
}) 
// imagine you want to submit some user action without blocking the response
// in this case, you can't rely on the user in the context to do some action in the middleware
```

I tried to solve this problem in my way with the help of your suggestion about extending `PrismaClient`: (it's not my final solution, it needs a bit of work to consider every edge case, but my point is to share my thoughts here)
```javascript

type DBContext = {
  user: any;
};

interface ContextAwarePrismaClient extends PrismaClient {
  $setContext: (context: DBContext) => ContextAwarePrismaClient;
}

function addContextProxy(model: Object, _context: DBContext) {
  return new Proxy(model, {
    get(target, p, receiver) {
      const method = Reflect.get(target, p, receiver);

      if (typeof method !== 'function') {
        return method;
      }

      return (args: unknown) => {
        if (typeof args !== 'object') {
          return method.call(target, args);
        }

        return method.call(target, { ...args, _context });
      };
    },
  });
}

function setContext(
  this: ContextAwarePrismaClient,
  _context: DBContext
): ContextAwarePrismaClient {
  return new Proxy(this, {
    get(target, p, receiver) {
      const original = Reflect.get(target, p, receiver);

      if (
        typeof p !== 'string' ||
        /^\$.+/.test(p) ||
        typeof original !== 'object'
      ) {
        return original;
      }

      return addContextProxy(original, _context);
    },
  });
}

const consumeContextMiddleware: Prisma.Middleware = async (params, next) => {
  const result = await next(omit(params, 'args._context')); // you need to remove _context before passing to next middleware
  
  return result;
};

function createPrismaClient() {
  const prisma = Object.create(new PrismaClient(), {
    $setContext: {
      value: setContext,
      enumerable: false,
      writable: false,
      configurable: false,
    },
  }) as ContextAwarePrismaClient;

  
  prisma.$use(consumeContextMiddleware);

  return prisma;
}
```

This way, you will get what you passed to the context for every query.

```javascript
prisma.setContext({user}).model.create({}) // only create knows about the context

const prismaWithGlobalContext = prisma.setContext({user})

prismaWithGlobalContext.model.create({}) 
prismaWithGlobalContext.model2.create({})  // all actions can access to the same context

```

**Reply by icyrizard** [reply]
@ali-shabani I've been using the Proxy approach now (with success), so thanks a lot for that! 

I struggle a little bit of wrapping my mind around what is really happening here, without knowing the internals of the `PrismaClient` I think I'll be forever in the dark until I make a study out of that. I am a bit worried what happens in future releases of Prisma in terms of stability. This approach might easily break. Maybe there are some people watching this thread with an opinion on that?

I do see the extension alternative but mentioned in this thread, but I find it difficult to apply to my exact use case (I want to insert the context and have it available in middleware.

**Anyhoo**, the reason I'm reaching out, I have a small addition to your solution @ali-shabani.

One minor improvement I wanted to share **(although really simple)** is to check for `p === 'transaction`, see code below. This prevents an error which I've also put below. (Beware that it a bit of the code, so not really **copy-pastable**)

~~~typescript
function addContextProxy(model: Object, _context: DBContext) {
  return new Proxy(model, {
    get(target, p, receiver) {
      const method = Reflect.get(target, p, receiver);

      //  ======= I just added the p === 'transaction' so it ignores when it's a transaction function call. ====
      if (typeof method !== 'function' || p === 'transaction') {
        return method;
      }

      return (args: unknown) => {
        if (typeof args !== 'object') {
          return method.call(target, args);
        }

        return method.call(target, { ...args, _context });
      };
    },
  });
}
~~~

Now, I'm not a 100% sure this is the right approach, it does solve the issue of an internal error being raised if not providing this. I will show how I use the code below. If not for the `p === 'transaction'` check, you will get an error like below: `error: Failed to convert napi `Undefined` into rust type String`

~~~bash
[1] { method: false, p: 'enabled' }
[1] { method: false, p: 'enabled' }
[1] { method: [AsyncFunction: transaction], p: 'transaction' }
[1] error: Failed to convert napi `Undefined` into rust type `String`
[1] Error: Failed to convert napi `Undefined` into rust type `String`
[1]     at Vt.transaction (.... /node_modules/@prisma/client/runtime/library.js:100:996)
[1]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[1]     at async Proxy._transactionWithCallback (... /node_modules/@prisma/client/runtime/library.js:174:9308)
....
....
~~~
 
## For completeness - here is my approach

See at the end of this code how I use this in a transaction. The scope will remain in all calls of the transaction `tx` variable (without the error shown above!).

~~~typescript
export function tenantScope(tenantId: number | null, query: Record<string, any>) {
    const where = {
        ...query?.where || {}
    };

    if (tenantId) {
        where.tenant = {
            id: tenantId,
        }
    }

    return {
        ...query,
        where: where,
    }
}

const ModelMap = {
    User: {
        tenantScoped: true,
        softDeleteScoped: true,
    }
}

/**
 * Initialize Prisma middleware.
 *
 * @param req
 * @private
 */
prismaClient.$use(async (params, next) => {
    if (ModelMap[params.model]) {
        if (['findMany', 'find', 'findFirst', 'findFirstOrThrow', 'count', 'deleteMany', 'delete'].includes(params.action)) {
            if (ModelMap[params.model].tenantScoped && params.args?._context?.tenantId) {
                params.args = tenantScope(params.args?._context?.tenantId, params.args)

            }
        }
    }

    return next(omit(params, 'args._context'))
});

const prisma = (context: ApiContext = null, rest = {}): ContextAwarePrismaClient => {
    if (context?.tenantId) {
        return prismaClient.$setContext({ ...rest, tenantId: context?.tenantId });
    } else {
        return prismaClient;
    }
}

const context = { tenantId: 1 };

// === the actual usage and the reason I made an adjustment ===
await prisma(context).$transaction(async (tx) => {
            const where = {
                id: {
                     in: [1, 2, 3]
                }
            };

            const count = await tx.users.count({ where });
            const users = await tx.users.findMany({where}});

           return [users, count]
})
~~~

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
